File: dsa-practice-ram/dsa-practice-ram/Arrays/MaxSubArraySum.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Text;
using System.Threading.Tasks;

namespace dsa_practice_ram.Arrays
{
    internal class MaxSubArraySum
    {
        public void MaxSubArraySumFixedSlidingWindow()
        {
            // This approach  works for both negative and non negative integers
            int[] array = { 1, 4, 5, 6, 1, 2, -3, 6, 8, -9, 7, 6, 3, 3 };
            int k = 3;

            Range range = ..k;
            int windowSum = 0;

            windowSum = array[..k].Sum();
            int maxSum = windowSum;

            for (int i = k; i < array.Length; i++)
            {
                windowSum += array[i] - array[i - k];
                if (windowSum > maxSum)
                {
                    range = new Range(i - k + 1, i + 1);
                    maxSum = windowSum;
                }
            }
            int[] subArray = array[range];
            Console.WriteLine($"{string.Join(", ", subArray)} -> {maxSum}");
        }

        public void KadanesAlgorithmGetSubArray()
        {
            // You restart the subarray from the current element only when the ongoing total (currentSum) isn’t worth carrying forward.
            // Return the contiguous subarray of any length that has maximum sum
            int[] nums = { -2, 1, -3, 4, -1, 2, 1, -5, 4 };

            int sum = 0;
            int maxSum = int.MinValue;
            Range range = ..1;
            int tempStart = 0;

            for (int i = 0; i < nums.Length; i++)
            {
                if (sum <= 0)
                {
                    sum = nums[i];
                    tempStart = i;
                }
                else
                {
                    sum += nums[i];
                }
                if (sum > maxSum)
                {
                    maxSum = sum;
                    range = tempStart..(i + 1);
                }
            }

            Console.WriteLine($"{string.Join(", ", nums[range])} -> {maxSum}");
        }

        public void KadanesAlgorithmGetMaxSumAnotherApproach()
        {
            // Given an array of integers (which may include negative numbers), 
            // find the contiguous sub array that has the maxumum sum
            int[] nums = { 2, -3, 4, -1, -2, 1, 5, -3 };
            int maxSum = int.MinValue;
            int sum = 0;

            foreach (var num in nums)
            {
                sum += num;
                maxSum = Math.Max(maxSum, sum);

                if (sum < 0)
                    sum = 0;
            }

            Console.WriteLine(maxSum);
        }

        public void KadanesAlgorithmGetMaxSumAnotherApproachSubArray()
        {
            // int[] nums = { 2, -3, 4, -1, -2, 1, 5, -3 };
            int[] nums = { -4, -3, -2, -1 };
            int maxSum = int.MinValue;
            int sum = 0;
            Range range = ..1;
            int tempStart = 0;

            for (int i = 0; i < nums.Length; i++)
            {
                sum += nums[i];

                if (sum > maxSum)
                {
                    maxSum = sum;
                    range = tempStart..(i + 1);
                }

                if (sum < 0)
                {
                    sum = 0;
                    tempStart = i + 1;
                }
            }

            Console.WriteLine($"{string.Join(", ", nums[range])} -> {maxSum}");
        }
    }
}

File: MyLearning/PythonPractice/dsa_practice/sliding_window/max_sub_array_sum.py
"""Maximum subarray sum utilities."""

INT_MIN = -2 ** 31

# C# MaxSubArraySum.MaxSubArraySumFixedSlidingWindow

def max_subarray_sum_fixed_sliding_window(array=None, k=3):
    """Return the max-sum subarray of size k and its sum.

    Problem: Find the fixed-size subarray with maximum sum.
    Approach: Sliding window with incremental sum updates.
    Time complexity: O(n)
    Space complexity: O(1)
    """
    if array is None:
        array = [1, 4, 5, 6, 1, 2, -3, 6, 8, -9, 7, 6, 3, 3]

    if not array or k <= 0 or k > len(array):
        return [], 0

    window_sum = sum(array[:k])
    max_sum = window_sum
    best_range = (0, k)

    for i in range(k, len(array)):
        window_sum += array[i] - array[i - k]
        if window_sum > max_sum:
            best_range = (i - k + 1, i + 1)
            max_sum = window_sum

    return list(array[best_range[0]:best_range[1]]), max_sum


# C# MaxSubArraySum.KadanesAlgorithmGetSubArray

def kadanes_algorithm_get_subarray(nums=None):
    """Return the max-sum contiguous subarray and its sum.

    Problem: Find the contiguous subarray with maximum sum.
    Approach: Kadane's algorithm tracking the best start/end.
    Time complexity: O(n)
    Space complexity: O(1)
    """
    if nums is None:
        nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]

    if not nums:
        return [], 0

    current_sum = 0
    max_sum = INT_MIN
    best_range = (0, 1)
    temp_start = 0

    for i, value in enumerate(nums):
        if current_sum <= 0:
            current_sum = value
            temp_start = i
        else:
            current_sum += value

        if current_sum > max_sum:
            max_sum = current_sum
            best_range = (temp_start, i + 1)

    return list(nums[best_range[0]:best_range[1]]), max_sum


# C# MaxSubArraySum.KadanesAlgorithmGetMaxSumAnotherApproach

def kadanes_algorithm_get_max_sum(nums=None):
    """Return the maximum subarray sum.

    Problem: Find the maximum sum of any contiguous subarray.
    Approach: Kadane's algorithm with running sum reset.
    Time complexity: O(n)
    Space complexity: O(1)
    """
    if nums is None:
        nums = [2, -3, 4, -1, -2, 1, 5, -3]

    if not nums:
        return 0

    max_sum = INT_MIN
    running_sum = 0

    for num in nums:
        running_sum += num
        if running_sum > max_sum:
            max_sum = running_sum

        if running_sum < 0:
            running_sum = 0

    return max_sum


# C# MaxSubArraySum.KadanesAlgorithmGetMaxSumAnotherApproachSubArray

def kadanes_algorithm_get_max_sum_subarray(nums=None):
    """Return the max-sum contiguous subarray and its sum.

    Problem: Find the max-sum subarray even when all values are negative.
    Approach: Kadane's algorithm with range tracking.
    Time complexity: O(n)
    Space complexity: O(1)
    """
    if nums is None:
        nums = [-1, -4, -2, -5]

    if not nums:
        return [], 0

    max_sum = INT_MIN
    running_sum = 0
    best_range = (0, 1)
    temp_start = 0

    for i, num in enumerate(nums):
        running_sum += num

        if running_sum > max_sum:
            max_sum = running_sum
            best_range = (temp_start, i + 1)

        if running_sum < 0:
            running_sum = 0
            temp_start = i + 1

    return list(nums[best_range[0]:best_range[1]]), max_sum
